langages/scala/programmer-en-scala-3-booleens-et-conditions.md

```diff
@@ -0,0 +1,277 @@
+
+Dans le monde réel, un programme informatique a souvent besoin de vérifier si certaines propositions sont vraies ou
+fausses. En reprenant l'exemple du jeu Mario Bros, le jeu doit vérifier si :
+
+- Mario est vivant
+- Mario possède encore des vies
+- Mario a 100 pièces
+- etc...
+
+Une proposition est représentée par un type de données très courant en informatique : un booléen. Ce type ne peut
+prendre que deux valeurs : "vrai" ou "faux".
+
+En Scala, le type "booléen" s'écrit `Boolean`. Les valeurs "vrai" et "faux" s'écrivent respectivement `true` et `false`.
+
+```scala
+val x = true // "vrai"
+val y = false // "faux"
+```
+
+### Opérations booléennes
+
+À l'instar des types de données vus dans le cours précédent, les booléens possèdent des opérations qui leur sont propres
+dont trois principales, "non", "ou", "et".
+
+*Note : ces opérations sont issues de l'[Algèbre de Boole](https://fr.wikipedia.org/wiki/Alg%C3%A8bre_de_Boole_(logique))
+, souvent étudié en classe préparatoire ou en licence de mathématiques.*
+
+#### Non
+
+L'opération "non" est une opération qui retourne "vrai" si le booléen donné est "faux" et "faux" si le booléen est "
+vrai". En pratique, on peut dire que cette opération "inverse" le booléen donné :
+
+| Entrée | Résultat |
+|--------|----------|
+| Vrai   | Faux     |
+| Faux   | Vrai     |
+
+Le "non" en Scala s'écrit `!` avant la valeur à "inverser" :
+
+```scala
+!true // false
+```
+
+#### Ou
+
+Le "ou" est une opération qui prend deux booléens pour renvoyer "vrai" si __au moins__ un des deux booléens l'est également. On parle aussi de "ou inclusif" :
+
+| A    | B    | Résultat |
+|------|------|----------|
+| Vrai | Vrai | Vrai     |
+| Vrai | Faux | Vrai     |
+| Faux | Vrai | Vrai     |
+| Faux | Faux | Faux     |
+
+Cette opération s'écrit `||` en Scala :
+
+```scala
+true || false // true
+```
+
+#### Et
+
+L'opération "et", à l'instar du "ou", prend deux booléens. Elle renvoie "vrai" si et seulement si les deux booléens sont
+vrais :
+
+| A    | B    | Résultat |
+|------|------|----------|
+| Vrai | Vrai | Vrai     |
+| Vrai | Faux | Faux     |
+| Faux | Vrai | Faux     |
+| Faux | Faux | Faux     |
+
+Cette opération s'écrit `&&`:
+
+```scala
+true && true // true
+```
+
+## Les comparaisons
+
+Hormis les opérations entre booléens citées ci-dessus, Scala met à disposition d'autres opérateurs qui retournent des
+booléens comme les comparaisons.
+
+### Égalité
+
+L'opérateur classique pour comparer deux valeurs en Scala est le double égal `==`:
+
+```scala
+1 == 2 // false
+4 == 4 // vrai
+"a" == "a" // vrai
+```
+
+Attention : comme vu dans le cours précédent, certaines valeurs, bien que similaires, n'ont pas le même type ! Elles ne
+sont donc pas égales :
+
+```scala
+"1" == 1 // false
+```
+
+Comme la négation de l'égalité ("est différent de") est une opération très courante, Scala comme beaucoup d'autres
```

Comment:
j'aurais plutot créé un titre ###Inégalité ici

Reply:
AFAIK, ce n'est techniquement pas une inégalité.

Selon [la page française de Wikipedia sur le sujet](https://fr.wikipedia.org/wiki/In%C3%A9galit%C3%A9_(math%C3%A9matiques)) :
> Contrairement à une interprétation [étymologique](https://fr.wikipedia.org/wiki/%C3%89tymologie), la [négation](https://fr.wikipedia.org/wiki/N%C3%A9gation_logique) d’une [égalité](https://fr.wikipedia.org/wiki/%C3%89galit%C3%A9_(math%C3%A9matiques)) (avec le symbole ≠) n’est pas considérée comme une inégalité[2](https://fr.wikipedia.org/wiki/In%C3%A9galit%C3%A9_(math%C3%A9matiques)#cite_note-2) et se traite différemment.